### PR TITLE
[Concurrency] Read global variable from function call

### PR DESCRIPTION
**Pull request by Anthonysdu**
There are missing context switch points when the arguments in a function call are global variables, leading to incorrect false. (See the test case.)

### CODE REVIEW

**Comment by Anthonysdu** [comment]
benchexec: https://github.com/esbmc/esbmc/actions/runs/12270411020

**Comment by Anthonysdu** [comment]
wait.. I wonder why the global var passed to that function didn't cause a context switch, should take a look before this pr.

**Comment by Anthonysdu** [comment]
> wait.. I wonder why the global var passed to that function didn't cause a context switch, should take a look before this pr.

```
thread id: 1, 471, file main.c line 5 column 3 function a
c:main.c@65@F@a@b
```
the symbol is the function parameter name and wasn't considered as global.

**Comment by lucasccordeiro** [comment]
> the symbol is the function parameter name and wasn't considered as global.

Shall we extract the function parameter name?

File: src/goto-symex/execution_state.cpp [hunk 1]
```diff
@@ -301,8 +301,73 @@ void execution_statet::symex_step(reachability_treet &art)
...
+        dereference(operand, dereferencet::READ);
+
+    // Always run intrinsics, whether guard is false or not. This is due to the
+    // unfortunate circumstance where a thread starts with false guard due to
+    // decision taken in another thread in this trace. In that case the
+    // terminate intrinsic _has_ to run, or we explode.
+    if (is_symbol2t(call.function))
+    {
+      const irep_idt &id = to_symbol2t(call.function).thename;
+      if (has_prefix(id.as_string(), "c:@F@__ESBMC"))
```
Comment (ChenfengWei1) [inline src/goto-symex/execution_state.cpp]: Does it also work in C++?

Comment (Anthonysdu) [reply src/goto-symex/execution_state.cpp]: This function is from it's parent class, I presume it should work.

But this PR is broken with some tests, will need to check it later.

**Comment by Anthonysdu** [comment]
> > the symbol is the function parameter name and wasn't considered as global.
> 
> Shall we extract the function parameter name?

The argument was simplified to a boolean value in symex_goto that's why we lost track of it.

File: src/goto-symex/execution_state.cpp (hunk 1 above)
Comment (ChenfengWei1) [reply src/goto-symex/execution_state.cpp]: In my experience, the function name in C++ might be different compared to C, but I am not sure if it’s the case for the intrinsics function.

File: src/goto-symex/symex_function.cpp
Comment (Anthonysdu) [inline src/goto-symex/symex_function.cpp:1]: It's werid that the comment above said read the arguments before renaming but it didn't actually read them.

File: src/goto-symex/symex_function.cpp [hunk 2]
```diff
@@ -251,7 +251,10 @@ void goto_symext::symex_function_call_code(const expr2tc &expr)
   // read the arguments -- before the locality renaming
```
Comment (Anthonysdu) [inline src/goto-symex/symex_function.cpp:251]: It's werid that the comment above said read the arguments before renaming but it didn't actually read them.

**Comment by Anthonysdu** [comment]
benchexec:https://github.com/esbmc/esbmc/actions/runs/12306951840

**Review by lucasccordeiro — APPROVED** [review]

**Comment by lucasccordeiro** [comment]
Thanks for submitting this PR, @Anthonysdu.

**Review by XLiZHI — APPROVED** [review]

**Pull request merged**